File: cpp/442. Find All Duplicates in an Array.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<int> vi;
typedef vector<vector<int>> vvi;
typedef pair<int, int> pi;

class Solution {
// Time: O(n), Space: O(1)
public:
    vector<int> findDuplicates(vector<int>& nums) {
        vi ans;
        for(int num : nums) {
            num = abs(num);
            int idx = num - 1;
            if(nums[idx] < 0) ans.push_back(num);
            else {
                nums[idx] = -nums[idx]; 
            }
        }      
        return ans;
    }
};

int main() {
    ios::sync_with_stdio(0); 
    cin.tie(0);

    Solution solution;
    vi nums = {4,3,2,7,8,2,3,1};
    vi ans = solution.findDuplicates(nums);
    assert(ans==vi({2, 3}));
}

File: python/442-Find All Duplicates in an Array.py
from typing import List

'''
Very tricky for Time: O(n) and Space: O(1) constraints
Key: 1 ≤ a[i] ≤ n, so we can use array index to mimic our hash table
https://www.youtube.com/watch?v=aMsSF1Il3IY
'''

class Solution:
    def findDuplicates(self, nums: List[int]) -> List[int]:
        "Time: O(n), Space: O(1)"
        result = []
        for num in nums:
            key = abs(num) - 1
            val = abs(num)
            if nums[key] < 0:
                result.append(val)
            else:
                nums[key] = -nums[key]
        return result

nums = [4,3,2,7,8,2,3,1]
result = Solution().findDuplicates(nums)
assert result == [2, 3]